### PR TITLE
Submit the marine/soca prep and run step as a batch job

### LINKED ISSUE #95: SOCA: Leverage run_jedi_exe.py to execute the prep and run scripts
## Description
As an alternative to running a single jedi executable, modify `run_jedi_exe.py` to submit a job that will execute `exgdas_global_marine_analysis_prep.py` and `exgdas_global_marine_analysis_run.sh` in sequence.

### LINKED ISSUE #97: SOCA bug: Harcoded date in stage.py
Oops:
https://github.com/NOAA-EMC/GDASApp/blob/5952c9df74b37d10f3aa2282f6fa36f287348efc/ush/ufsda/stage.py#L93

### PR DESCRIPTION
**Pull request by guillaumevernieres**
## Description
Leverage the `run_jedi_exe.py` application to allow submission of a batch script that will execute:
```
 exgdas_global_marine_analysis_prep.py
```
and
```
exgdas_global_marine_analysis_run.sh
```

### Features added and bug fixes
- removed hard-coded date ... Oops
- generate MOM6 /SOCA Fortran namelist from configuration
- pass DA variables from config ... not passed down to the yamls yet 
- I got tiered of `chmoding +x`  the `submit_job.sh` bash script generated by `run_jedi_exe.py`. It is now made executable when written by `run_jedi_exe.py`

### Known problems
- I made assumptions on what the `global-workflow` (GW) environment is, it's probably mostly wrong but it will be fixed once we have access to the UFS-mom6/cic6/data atm model as a cycling option of the GW ... Patience!
- The added feature is currently not tested as part of the `ci`. To be revisited in the future and possibly added as part of a suite or **regression tests**.
- I have not tested the atmospheric application to make sure I did not break that part of the code ... TODO before I un-draft this PR
- It definitely will not work as is for the atmospheric `prep` and `run`. That need to be sorted out once I understand better the GW environement
- I need to add the R2D2 config to the experiment yaml, it currently can only point to a tiny R2D2 test database 

### How to test this PR on Orion
I have only prepared static files for `Orion` so far. Here's what needs to be done to test:

```console
$ git clone -b feature/submit-prep-run https://github.com/NOAA-EMC/GDASApp.git
$ cd GDASApp
$ module use $PWD/modulefiles
$ module load GDAS/orion
$ ./build.sh
```
it takes a while to build the entire bundle ... So better forget for an hour and go do something else.

There are many ways one could prepare an experiment, here's one:

 #### Create an experiment directory somewhere
```console
$ mkdir test-soca-gdasapp     # Create an experiment directory somewhere
$ cd test-soca-gdasapp
$ ln -s <pat to GDASApp>/ush/run_jedi_exe.py . # link of convenience
$ cp <pat to GDASApp>/ush/examples/run_scripts/soca5deg.yaml .  # make a copy of an example yaml config
```
#### Edit `soca5deg.yaml` 
```yaml
working directory: /work2/noaa/da/gvernier/runs/test-gdasapp/test-5deg # Needs editing
# replace the above by working directory: <path to test-soca-gdasapp>/test-5deg
GDASApp home: /work2/noaa/da/gvernier/GDASApp    # replace by yours
GDASApp mode: gw_scripts
GDASApp bin: /work2/noaa/da/gvernier/GDASApp/build/bin  # replace by yours

executable options:
  obs_yaml_dir: /work2/noaa/da/gvernier/GDASApp/parm/soca/obs/config  # edit path
  obs_list: /work2/noaa/da/gvernier/GDASApp/parm/soca/obs/obs_list.yaml   # edit path
  atm_window_length: PT6H
  valid_time: 2018-04-15T12:00:00Z
  dump: gdas
  case: C768
  levs: 128

job options:
  machine: orion
  account: da-cpu
  queue: debug
  partition: debug
  walltime: '30:00'
  ntasks: 6
  modulepath: /work2/noaa/da/gvernier/GDASApp/modulefiles   # edit path
  mpiexec: mpirun
  mpinproc: '-np'

jedi static:
  soca:
    path: /work2/noaa/da/gvernier/data/static/72x35x25/soca       # where the static files are, do not edit that line
    variables: tocn,socn,ssh

model background:
    ocn: /work2/noaa/da/gvernier/data/test_bkg/72x35x25/       # do not edit for now

r2d2 options:
    root: /work2/noaa/da/gvernier/data/static/r2d2-shared        # do not edit for now

fms:
    domain_stack_size: 5832000
```
#### Run the application
```console
$ ./run_jedi_exe.py -c  soca5deg.yaml
```
this will submit a batch job that will run the `soca` prep and run scripts.

#### Monitor 
The output should be under `./test-5deg/GDASApp.o*`

## Issues addressed
- fixes #97 
- fixes #95

### CODE REVIEW

**Comment by guillaumevernieres** [comment]
I have just tested one of the atmospheric application (`3dhofx_orion.yaml`) here:
/work2/noaa/da/gvernier/runs/test-gdasapp/gdas_single_test_hofx3d

it completed successfully. :tada: ... marking it as ready to review.

**Review by liuxiao37k — DISMISSED** [review]
I ran this workflow as a black box. It indeed successfully submitted a batch job and ran to completion (I think) with an exit message:
```
Run: Finishing oops::Variational<SOCA, UFO and IODA observations> with status = 0
OOPS Ending   2022-07-25 11:40:30 (UTC-0500)
+ set +x
Mon Jul 25 11:40:30 CDT 2022 EXITING /work/noaa/ng-godas/xliu/GDASApp/modulefiles/../scripts/exgdas_global_marine_analysis_run.sh with return code
```
The instruction is current and easy to follow. No issues encountered (I just needed to remind myself of starting off with a clean terminal due to potential conflicts with pre-loaded modules.)

Approving for now and will revisit for a better understanding.

File: ush/ufsda/disk_utils.py [hunk 1]
```diff
@@ -16,6 +17,41 @@ def mkdir(dirpath):
         logging.info(f"{dirpath} could not be created")
 
 
+def removefile(file):
```
Comment (CoryMartin-NOAA) [inline ush/ufsda/disk_utils.py:20]: Why are some of these shutil and then the first one just `os.remove`?

Comment (CoryMartin-NOAA) [reply ush/ufsda/disk_utils.py:20]: Lines 49-50 should be identical to 24-25, no?

File: ush/ufsda/misc_utils.py [hunk 2]
```diff
@@ -80,9 +81,25 @@ def create_batch_job(job_config, working_dir, executable, yaml_path):
...
+            scripts_path = os.path.join(job_config['modulepath'], '../', 'scripts')
+            # run the pre/run/post scripts
+            f.write(f"source load_envar.sh\n")
+            f.write(f"{scripts_path}/exgdas_global_marine_analysis_prep.py\n")
+            f.write(f"cd {working_dir}/analysis\n")
+            f.write(f"export APRUN_SOCAANAL=\"{job_config['mpiexec']} {job_config['mpinproc']} {job_config['ntasks']}\"\n")
+            f.write(f"{scripts_path}/exgdas_global_marine_analysis_run.sh\n")
     logging.info(f"Wrote batch submission script to {batch_script}")
+
+    # make the batch script executable
```
Comment (CoryMartin-NOAA) [inline ush/ufsda/misc_utils.py]: Why is this needed? Do you not just do `sbatch script.sh`?

File: ush/ufsda/misc_utils.py [hunk 3]
```diff
@@ -80,9 +81,25 @@ def create_batch_job(job_config, working_dir, executable, yaml_path):
...
 """
         f.write(commands)
-        if scheduler[job_config['machine']] == 'slurm':
-            f.write(f"srun -n $SLURM_NTASKS {executable} {yaml_path}\n")
+        if single_exec:
+            # standalone jedi application
+            if scheduler[job_config['machine']] == 'slurm':
+                f.write(f"srun -n $SLURM_NTASKS {executable} {yaml_path}\n")
+        else:
+            # TODO (Guillaume): Hard coded for soca in a few places, change that
```
Comment (CoryMartin-NOAA) [inline ush/ufsda/misc_utils.py:89]: Just making a note to resolve this sooner rather than later.

File: ush/run_jedi_exe.py [hunk 4]
```diff
@@ -47,67 +67,137 @@ def run_jedi_exe(yamlconfig):
...
-    ufsda.stage.gdas_fix(gdasfix, workdir, var_config)
-    # link executable
-    baseexe = os.path.basename(executable_subconfig['executable'])
-    ufsda.disk_utils.symlink(executable_subconfig['executable'], os.path.join(workdir, baseexe))
-    # create output directories
-    ufsda.disk_utils.mkdir(os.path.join(workdir, 'diags'))
-    if app_mode in ['variational']:
-        ufsda.disk_utils.mkdir(os.path.join(workdir, 'anl'))
-        ufsda.disk_utils.mkdir(os.path.join(workdir, 'bc'))
+    pdy = prev_cycle.strftime("%Y%m%d")
```
Comment (RussTreadon-NOAA) [inline ush/run_jedi_exe.py]: Normally `pdy` is the current cycle.   If we want `pdy` to hold the previous cycle PDY, we should rename this variable `gpdy` to be consistent with `gcyc` and `gdate`.

File: ush/ufsda/disk_utils.py (hunk 1 above)
Comment (guillaumevernieres) [reply ush/ufsda/disk_utils.py:20]: good catch @CoryMartin-NOAA ... and `shutil.rm` doesn't exist ...

File: ush/ufsda/misc_utils.py (hunk 2 above)
Comment (guillaumevernieres) [reply ush/ufsda/misc_utils.py]: I want to be able to do `./submit.sh`. Useful when using interactive job or debugging the prep step.

Comment (CoryMartin-NOAA) [reply ush/ufsda/misc_utils.py]: so this is to just debug and doesn't submit the job, but just runs like a standard `.sh` script?

**Comment by ShastriPaturi** [comment]
I was able to follow the instructions and run on orion.

File: ush/ufsda/misc_utils.py (hunk 2 above)
Comment (guillaumevernieres) [reply ush/ufsda/misc_utils.py]: yes

File: ush/run_jedi_exe.py (hunk 4 above)
Comment (guillaumevernieres) [reply ush/run_jedi_exe.py]: Noted, will make a change @RussTreadon-NOAA .

File: ush/ufsda/misc_utils.py (hunk 3 above)
Comment (guillaumevernieres) [reply ush/ufsda/misc_utils.py:89]: @CoryMartin-NOAA  On my todo list as soon as I can cycle mom6-cice6 in the global-workflow.

**Review by ShastriPaturi — DISMISSED** [review]

**Comment by RussTreadon-NOAA** [comment]
[Lines 112-116](https://github.com/NOAA-EMC/GDASApp/blob/feature/submit-prep-run/scripts/exgdas_global_marine_analysis_run.sh#L113-L116) of [scripts/exgdas_global_marine_analysis_run.sh](https://github.com/NOAA-EMC/GDASApp/blob/feature/submit-prep-run/scripts/exgdas_global_marine_analysis_run.sh) return an error code, `err`.  This is good, but `err` is not assigned a value in the script.   

This section of `exgdas_global_marine_analysis_run.sh` is not altered in this PR so it's not necessary to address this now.  At some point we should set and update `err` depending on the steps within the script for which we want error tracing.

**Review by RussTreadon-NOAA — COMMENTED** [review]
Cloned `feature/submit-prep-run` and built apps on Orion.  Successfully ran `soca5deg.yaml`.  Also successfully ran `3dhofx_orion.yaml` and `4dhofx_orion.yaml`.   `3dvar_orion.yaml` failed with the job exceeding the specified 10 minute job wall clock limit.   The hofx and 3dvar jobs are not part of this PR.   Apart from resolution of a few comments, this PR is ready to merge.

**Comment by guillaumevernieres** [comment]
> [Lines 112-116](https://github.com/NOAA-EMC/GDASApp/blob/feature/submit-prep-run/scripts/exgdas_global_marine_analysis_run.sh#L113-L116) of [scripts/exgdas_global_marine_analysis_run.sh](https://github.com/NOAA-EMC/GDASApp/blob/feature/submit-prep-run/scripts/exgdas_global_marine_analysis_run.sh) return an error code, `err`. This is good, but `err` is not assigned a value in the script.
> 
> This section of `exgdas_global_marine_analysis_run.sh` is not altered in this PR so it's not necessary to address this now. At some point we should set and update `err` depending on the steps within the script for which we want error tracing.

Thanks @RussTreadon-NOAA , will reference your comment in a new issue

**Comment by guillaumevernieres** [comment]
> Cloned `feature/submit-prep-run` and built apps on Orion. Successfully ran `soca5deg.yaml`. Also successfully ran `3dhofx_orion.yaml` and `4dhofx_orion.yaml`. `3dvar_orion.yaml` failed with the job exceeding the specified 10 minute job wall clock limit. The hofx and 3dvar jobs are not part of this PR. Apart from resolution of a few comments, this PR is ready to merge.

Thanks for your review @RussTreadon-NOAA . Can you click `approve` if you'r OK with the changes?

**Review by RussTreadon-NOAA — APPROVED** [review]
While some comments remain unresolved, this PR achieves the stated purpose.  Approve.

File: ush/ufsda/disk_utils.py (hunk 1 above)
Comment (guillaumevernieres) [reply ush/ufsda/disk_utils.py:20]: @CoryMartin-NOAA , made the changes you suggested

File: ush/ufsda/misc_utils.py (hunk 2 above)
Comment (guillaumevernieres) [reply ush/ufsda/misc_utils.py]: @CoryMartin-NOAA I removed the part that makes the script executable

Comment (guillaumevernieres) [reply ush/ufsda/misc_utils.py]: @CoryMartin-NOAA , referenced in issue #100

File: ush/run_jedi_exe.py (hunk 4 above)
Comment (guillaumevernieres) [reply ush/run_jedi_exe.py]: @RussTreadon-NOAA changed `pdy` as suggested.

File: ush/ufsda/misc_utils.py (hunk 3 above)
Comment (guillaumevernieres) [reply ush/ufsda/misc_utils.py:89]: see issue #100

File: ush/run_jedi_exe.py (hunk 4 above)
Comment (RussTreadon-NOAA) [reply ush/run_jedi_exe.py]: Thank you, @guillaumevernieres.   `pdy` goes with `valid_time`.  Commit at [49f8337](https://github.com/NOAA-EMC/GDASApp/commit/49f83372b4d3ca824ce10ea1b68933ee1409ef11) looks good.

**Review by CoryMartin-NOAA — APPROVED** [review]
